optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A> joiner1, BiJoiner<A, A> joiner2) {
+        return joinOther(otherClass, new BiJoiner[] {joiner1, joiner2});
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A> joiner1, BiJoiner<A, A> joiner2,
+            BiJoiner<A, A> joiner3) {
+        return joinOther(otherClass, new BiJoiner[] {joiner1, joiner2, joiner3});
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @param joiner4 never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A> joiner1, BiJoiner<A, A> joiner2,
+            BiJoiner<A, A> joiner3, BiJoiner<A, A> joiner4) {
+        return joinOther(otherClass, new BiJoiner[] {joiner1, joiner2, joiner3, joiner4});
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * <p>
+     * This method causes <i>Unchecked generics array creation for varargs parameter</i> warnings,
+     * but we can't fix it with a {@link SafeVarargs} annotation because it's an interface method.
+     * Therefore, there are overloaded methods with up to 4 {@link BiJoiner} parameters.
+     * @param otherClass never null
+     * @param joiners never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A>... joiners) {
+        BiJoiner<A, A> otherness = Joiners.notEqual();
+        BiJoiner[] allJoiners = Stream.concat(Stream.of(otherness), Arrays.stream(joiners))
+                .toArray(BiJoiner[]::new);
+        return join(otherClass, allJoiners);
+    }
+
+    // ************************************************************************
+    // If Exists
+    // ************************************************************************
+
+    /**
+     * Create a new {@link UniConstraintStream} for every A, if B exists.
+     * @param otherClass never null
+     * @param <B> the type of the second matched fact
+     * @return a stream that matches every A where a B exists
+     */
+    default <B> UniConstraintStream<A> ifExists(Class<B> otherClass) {
```

Comment:
Is this method every going to be useful?
What is the specification? If we have A1, A2, A3, ... they will match (= survive) if in these cases:
- there is a B1
- there is a B1, B2
but not if there not a single B.

Theory: "Nobody's every going to need this method." When in doubt, leave it out. Let the users tell us that they need and we 'll add it. Or let us feel the need for it in an example and let's then add it. But meanwhile, let's leave it out.

Reply:
I don't mind if we leave it out. I just remember having "meta-objects" in Drools working memory to signify some sort of a state change - do something if the state is X, but don't do it when it isn't.